pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,363 @@
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const clusterScopedNamespace = ""
+
+// placementResolution stores the selected objects and destinations for a single placement.
+// The mutex should be read locked before reading, and write locked before writing to any field.
+type placementResolution struct {
+	sync.RWMutex
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKeyMap map[string]*util.Key
+	destinations             []string
+}
+
+// updateObject adds/deletes an object to/from the resolution.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
+func (resolution *placementResolution) updateObject(obj runtime.Object) (bool, error) {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	key, err := util.KeyForGroupVersionKindNamespaceName(obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get key for object: %w", err)
+	}
+
+	// formatted-key to use for mapping
+	formattedKey := key.GvkNamespacedNameKey()
+	_, exists := resolution.objectIdentifierToKeyMap[formattedKey]
+
+	// avoid further processing for keys of objects being deleted that do not have a deleted object
+	if isBeingDeleted(obj) || key.DeletedObject != nil {
+		delete(resolution.objectIdentifierToKeyMap, formattedKey)
+
+		return exists, nil
+	}
+
+	// add object to map
+	resolution.objectIdentifierToKeyMap[formattedKey] = &key
+	// if an object is reconciled all the way into here, it means something internal changed, so we need to trigger
+	// a placement-decision update event in order for transport watchers to re-evaluate it
+	return true, nil
+}
+
+// removeObject deletes an object from the resolution if it exists.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
+func (resolution *placementResolution) removeObject(obj runtime.Object) bool {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	key, err := util.KeyForGroupVersionKindNamespaceName(obj)
+	if err != nil {
+		return false // assume object was never added
+	}
+
+	// formatted-key to use for mapping
+	formattedKey := key.GvkNamespacedNameKey()
+	_, exists := resolution.objectIdentifierToKeyMap[formattedKey]
+
+	delete(resolution.objectIdentifierToKeyMap, formattedKey)
+	return exists
+}
+
+// updateDestination updates the destinations list in the resolution.
+// This function is thread-safe.
+func (resolution *placementResolution) updateDestinations(destinations []string) {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	resolution.destinations = destinations
+}
+
+// toPlacementDecisionSpec converts the resolution to a placement decision
+// spec. This function is thread-safe.
+func (resolution *placementResolution) toPlacementDecisionSpec(gvkGvrMapper util.GvkGvrMapper) (*v1alpha1.PlacementDecisionSpec, error) {
+	resolution.RLock()
+	defer resolution.RUnlock()
+
+	workload := v1alpha1.DownsyncObjectReferences{}
```

Comment:
I do not see anything setting the `WorkloadGeneration` field.

Reply:
WIP, I think I should address it properly in a follow up PR. It came AFTER the majority of the code was written, and it requires some refactoring in order for it not to just be set on every update.

Reply:
@MikeSpreitzer actually, for correctness sake, I will add a simple implementation that tracks generations but makes them local to the controller, and not the CR (which can be argued is correct anyway) - what do you think about that?

Reply:
Please open an issue for it and update #1692

Reply:
Implemented [here](https://github.com/kubestellar/kubestellar/pull/1681/commits/6218a3c7691582cf2ebe17cd2636daa3391ea5e2).

Note that this makes the generation local to the controller, which is unavoidable unless we want to sync state on load (fill the resolutions with the content of the existing placement-decisions, before syncing objects). I'll open an issue.

Reply:
I am not clear on what you mean by "while the controller stabilizes" and "sync generation/content on load". Let's pursue those in review of that code.